Pypy support

@tiangolo asked about pypy support. @messense do you have input?

Looks like pyo3 does support pypy, see [here](https://pyo3.rs/v0.16.4/) and https://github.com/PyO3/rust-numpy/issues/219.

But I know pydantic-core uses some non abi3 methods, maybe we use stuff that would cause problems with pypy, we should probably try it sooner rather than later.

Well, I've had a go, see #159. 

The only feature which pydantic-core uses and isn't available in pypy is `get_item_unchecked` for list and tuple. I've resolved that with `pyo3-build-config`.

Two remaining problems:

Build on my m1 mac works, but when I try to run tests I get

```
ImportError while loading conftest '/Users/samuel/code/pydantic-core/tests/conftest.py'.
tests/conftest.py:11: in <module>
    from pydantic_core import SchemaValidator
pydantic_core/__init__.py:1: in <module>
    from ._pydantic_core import SchemaError, SchemaValidator, ValidationError, __version__
E   ImportError: dlopen(/Users/samuel/code/pydantic-core/pydantic_core/_pydantic_core.pypy39-pp73-darwin.so, 0x0006): tried: '/Users/samuel/code/pydantic-core/pydantic_core/_pydantic_core.pypy39-pp73-darwin.so' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64'))
```

Build on CI is passing and tests are running however, pypy is crashing with:

```
Fatal Python error: Aborted
Stack (most recent call first, approximate line numbers):
  File "/home/runner/work/pydantic-core/pydantic-core/tests/validators/test_typed_dict.py", line 787 in test_from_attributes_extra
```

See https://github.com/samuelcolvin/pydantic-core/runs/7253126026?check_suite_focus=true

Some other tests are failing, but I can't get the traceback due to the fatal error.

Any ideas? Hard for me to debug for the next few weeks until I get back to my desktop of x86.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -10,7 +10,7 @@ on:
 
 jobs:
   test:
-    name: test py${{ matrix.python-version }} rust ${{ matrix.rust-toolchain }}
+    name: test ${{ matrix.python-version }} rust ${{ matrix.rust-toolchain }}
     strategy:
       fail-fast: false
       matrix:

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -231,8 +231,8 @@ jobs:
   build:
     name: build on ${{ matrix.platform || matrix.os }} (${{ matrix.target }} - ${{ matrix.manylinux || 'auto' }})
     # only run on push to main and on release
-    if: "success() && (startsWith(github.ref, 'refs/tags/') || github.ref == 'refs/heads/main')"
-    needs: [test, lint]
+#    if: "success() && (startsWith(github.ref, 'refs/tags/') || github.ref == 'refs/heads/main')"
+#    needs: [test, lint]
     strategy:
       fail-fast: false
       matrix:

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -326,6 +326,7 @@ jobs:
           path: dist
 
       - run: ls -lh dist/
+      - run: echo "`ls src | wc -l` files"
 
   release:
     needs: [build, build-wasm-emscripten]

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -22,6 +22,9 @@ jobs:
           - '3.9'
           - '3.10'
           - '3.11.0-alpha - 3.11'
+          - 'pypy3.7'
+          - 'pypy3.8'
+          - 'pypy3.9'
         include:
           - rust-toolchain: stable
             python-version: '3.10'

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -231,8 +231,8 @@ jobs:
   build:
     name: build on ${{ matrix.platform || matrix.os }} (${{ matrix.target }} - ${{ matrix.manylinux || 'auto' }})
     # only run on push to main and on release
-#    if: "success() && (startsWith(github.ref, 'refs/tags/') || github.ref == 'refs/heads/main')"
-#    needs: [test, lint]
+    if: "success() && (startsWith(github.ref, 'refs/tags/') || github.ref == 'refs/heads/main')"
+    needs: [test, lint]
     strategy:
       fail-fast: false
       matrix:
@@ -332,7 +332,7 @@ jobs:
           path: dist
 
       - run: ls -lh dist/
-      - run: echo "`ls src | wc -l` files"
+      - run: echo "`ls dist | wc -l` files"
 
   release:
     needs: [build, build-wasm-emscripten]

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -244,10 +244,15 @@ jobs:
             platform: linux
           - os: windows
             ls: dir
+            interpreter: 3.7 3.8 3.9 3.10
           - os: windows
             ls: dir
             target: i686
             python-architecture: x86
+            interpreter: 3.7 3.8 3.9 3.10
+          - os: macos
+            target: aarch64
+            interpreter: 3.7 3.8 3.9 3.10
           - os: ubuntu
             platform: linux
             target: i686
@@ -261,6 +266,7 @@ jobs:
             platform: linux
             target: armv7
             container: messense/manylinux_2_24-cross:armv7
+            interpreter: 3.7 3.8 3.9 3.10
           # musllinux
           - os: ubuntu
             platform: linux
@@ -301,7 +307,7 @@ jobs:
           target: ${{ matrix.target }}
           manylinux: ${{ matrix.manylinux || 'auto' }}
           container: ${{ matrix.container }}
-          args: --release --out dist --interpreter 3.7 3.8 3.9 3.10 pypy3.7 pypy3.8 pypy3.9
+          args: --release --out dist --interpreter ${{ matrix.interpreter || '3.7 3.8 3.9 3.10 pypy3.7 pypy3.8 pypy3.9' }}
 
       - run: ${{ matrix.ls || 'ls -lh' }} dist/
 

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -301,7 +301,7 @@ jobs:
           target: ${{ matrix.target }}
           manylinux: ${{ matrix.manylinux || 'auto' }}
           container: ${{ matrix.container }}
-          args: --release --out dist --interpreter 3.7 3.8 3.9 3.10
+          args: --release --out dist --interpreter 3.7 3.8 3.9 3.10 pypy3.7 pypy3.8 pypy3.9
 
       - run: ${{ matrix.ls || 'ls -lh' }} dist/
 

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -41,3 +41,5 @@ codegen-units = 1
 
 [build-dependencies]
 version_check = "0.9.4"
+# used where logic has to be version/distrobution specific, e.g. pypy
+pyo3-build-config = "0.16.5"

--- a/build.rs
+++ b/build.rs
@@ -1,4 +1,5 @@
 fn main() {
+    pyo3_build_config::use_pyo3_cfgs();
     if let Some(true) = version_check::supports_feature("no_coverage") {
         println!("cargo:rustc-cfg=has_no_coverage");
     }

--- a/src/input/input_python.rs
+++ b/src/input/input_python.rs
@@ -462,15 +462,15 @@ fn mapping_seq_as_dict(seq: &PySequence) -> PyResult<&PyDict> {
     for r in seq.iter()? {
         let t: &PyTuple = r?.extract()?;
 
+        if t.len() != 2 {
+            return Err(PyTypeError::new_err("mapping items must be a tuple with 2 elements"));
+        }
+
         #[cfg(PyPy)]
         let k = t.get_item(0)?;
         #[cfg(PyPy)]
         let v = t.get_item(1)?;
 
-        #[cfg(not(PyPy))]
-        if t.len() != 2 {
-            return Err(PyTypeError::new_err("mapping items must be a tuple with 2 elements"));
-        }
         #[cfg(not(PyPy))]
         let k = unsafe { t.get_item_unchecked(0) };
         #[cfg(not(PyPy))]

--- a/src/input/input_python.rs
+++ b/src/input/input_python.rs
@@ -461,18 +461,20 @@ fn mapping_seq_as_dict(seq: &PySequence) -> PyResult<&PyDict> {
     let dict = PyDict::new(seq.py());
     for r in seq.iter()? {
         let t: &PyTuple = r?.extract()?;
-        let k: &PyAny;
-        let v: &PyAny;
-        if cfg!(PyPy) {
-            k = t.get_item(0)?;
-            v = t.get_item(1)?;
-        } else {
-            if t.len() != 2 {
-                return Err(PyTypeError::new_err("mapping items must be a tuple with 2 elements"));
-            }
-            k = unsafe { t.get_item_unchecked(0) };
-            v = unsafe { t.get_item_unchecked(1) };
+
+        #[cfg(PyPy)]
+        let k = t.get_item(0)?;
+        #[cfg(PyPy)]
+        let v = t.get_item(1)?;
+
+        #[cfg(not(PyPy))]
+        if t.len() != 2 {
+            return Err(PyTypeError::new_err("mapping items must be a tuple with 2 elements"));
         }
+        #[cfg(not(PyPy))]
+        let k = unsafe { t.get_item_unchecked(0) };
+        #[cfg(not(PyPy))]
+        let v = unsafe { t.get_item_unchecked(1) };
         dict.set_item(k, v)?;
     }
     Ok(dict)

--- a/src/input/input_python.rs
+++ b/src/input/input_python.rs
@@ -461,11 +461,18 @@ fn mapping_seq_as_dict(seq: &PySequence) -> PyResult<&PyDict> {
     let dict = PyDict::new(seq.py());
     for r in seq.iter()? {
         let t: &PyTuple = r?.extract()?;
-        if t.len() != 2 {
-            return Err(PyTypeError::new_err("mapping items must be a tuple with 2 elements"));
+        let k: &PyAny;
+        let v: &PyAny;
+        if cfg!(PyPy) {
+            k = t.get_item(0)?;
+            v = t.get_item(1)?;
+        } else {
+            if t.len() != 2 {
+                return Err(PyTypeError::new_err("mapping items must be a tuple with 2 elements"));
+            }
+            k = unsafe { t.get_item_unchecked(0) };
+            v = unsafe { t.get_item_unchecked(1) };
         }
-        let k = unsafe { t.get_item_unchecked(0) };
-        let v = unsafe { t.get_item_unchecked(1) };
         dict.set_item(k, v)?;
     }
     Ok(dict)

--- a/src/validators/mod.rs
+++ b/src/validators/mod.rs
@@ -427,7 +427,11 @@ pub struct BuildContext {
     depth: usize,
 }
 
+#[cfg(not(PyPy))]
 const MAX_DEPTH: usize = 100;
+
+#[cfg(PyPy)]
+const MAX_DEPTH: usize = 50;
 
 impl BuildContext {
     pub fn prepare_slot(&mut self, slot_ref: String) -> PyResult<usize> {

--- a/src/validators/recursive.rs
+++ b/src/validators/recursive.rs
@@ -118,7 +118,11 @@ impl Validator for RecursiveRefValidator {
 // see #143 this is a backup in case the identity check recursion guard fails
 // if a single validator "depth" (how many times it's called inside itself) exceeds the limit,
 // we raise a recursion error.
+#[cfg(not(PyPy))]
 const BACKUP_GUARD_LIMIT: u16 = 255;
+
+#[cfg(PyPy)]
+const BACKUP_GUARD_LIMIT: u16 = 123;
 
 fn guard_validate<'s, 'data>(
     validator_id: usize,

--- a/src/validators/typed_dict.rs
+++ b/src/validators/typed_dict.rs
@@ -677,11 +677,10 @@ impl<'a> Iterator for IterAttributes<'a> {
         // or we get to the end of the list of attributes
         loop {
             if self.index < self.attributes.len() {
-                let name: &PyAny = if cfg!(PyPy) {
-                    self.attributes.get_item(self.index).unwrap()
-                } else {
-                    unsafe { self.attributes.get_item_unchecked(self.index) }
-                };
+                #[cfg(PyPy)]
+                let name: &PyAny = self.attributes.get_item(self.index).unwrap();
+                #[cfg(not(PyPy))]
+                let name: &PyAny = unsafe { self.attributes.get_item_unchecked(self.index) };
                 self.index += 1;
                 // from benchmarks this is 14x faster than using the python `startswith` method
                 let name_cow = name

--- a/src/validators/typed_dict.rs
+++ b/src/validators/typed_dict.rs
@@ -677,7 +677,11 @@ impl<'a> Iterator for IterAttributes<'a> {
         // or we get to the end of the list of attributes
         loop {
             if self.index < self.attributes.len() {
-                let name: &PyAny = unsafe { self.attributes.get_item_unchecked(self.index) };
+                let name: &PyAny = if cfg!(PyPy) {
+                    self.attributes.get_item(self.index).unwrap()
+                } else {
+                    unsafe { self.attributes.get_item_unchecked(self.index) }
+                };
                 self.index += 1;
                 // from benchmarks this is 14x faster than using the python `startswith` method
                 let name_cow = name

--- a/src/validators/typed_dict.rs
+++ b/src/validators/typed_dict.rs
@@ -1,7 +1,7 @@
 use pyo3::exceptions::{PyAttributeError, PyTypeError};
-use pyo3::{intern, PyTypeInfo};
 use pyo3::prelude::*;
 use pyo3::types::{PyDict, PyFunction, PyList, PySet, PyString};
+use pyo3::{intern, PyTypeInfo};
 
 use ahash::AHashSet;
 
@@ -693,10 +693,19 @@ impl<'a> Iterator for IterAttributes<'a> {
                         // we don't want bound methods to be included, is there a better way to check?
                         // ref https://stackoverflow.com/a/18955425/949890
                         let is_bound = matches!(attr.hasattr(intern!(attr.py(), "__self__")), Ok(true));
-                        // the is_instance_of::<PyFunction> catches `staticmethod`, but also any other function,
+                        // the PyFunction::is_type_of(attr) catches `staticmethod`, but also any other function,
                         // I think that's better than including static methods in the yielded attributes,
                         // if someone really wants fields, they can use an explicit field, or a function to modify input
                         if !is_bound && !PyFunction::is_type_of(attr) {
+                            // MASSIVE HACK! PyFunction::is_type_of(attr) doesn't detect staticmethod on PyPy,
+                            // is_instance_of::<PyFunction> crashes with a null pointer, hence this hack, see
+                            // https://github.com/samuelcolvin/pydantic-core/pull/161#discussion_r917257635
+                            #[cfg(PyPy)]
+                            if attr.get_type().to_string() != "<class 'function'>" {
+                                return Some((name, attr));
+                            }
+
+                            #[cfg(not(PyPy))]
                             return Some((name, attr));
                         }
                     }

--- a/src/validators/typed_dict.rs
+++ b/src/validators/typed_dict.rs
@@ -1,5 +1,5 @@
 use pyo3::exceptions::{PyAttributeError, PyTypeError};
-use pyo3::intern;
+use pyo3::{intern, PyTypeInfo};
 use pyo3::prelude::*;
 use pyo3::types::{PyDict, PyFunction, PyList, PySet, PyString};
 
@@ -696,7 +696,7 @@ impl<'a> Iterator for IterAttributes<'a> {
                         // the is_instance_of::<PyFunction> catches `staticmethod`, but also any other function,
                         // I think that's better than including static methods in the yielded attributes,
                         // if someone really wants fields, they can use an explicit field, or a function to modify input
-                        if !is_bound && !matches!(attr.is_instance_of::<PyFunction>(), Ok(true)) {
+                        if !is_bound && !PyFunction::is_type_of(attr) {
                             return Some((name, attr));
                         }
                     }

--- a/tests/benchmarks/test_micro_benchmarks.py
+++ b/tests/benchmarks/test_micro_benchmarks.py
@@ -3,6 +3,7 @@ Numerous benchmarks of specific functionality.
 """
 import json
 import os
+import platform
 from datetime import date, datetime, timedelta, timezone
 from typing import Dict, FrozenSet, List, Optional, Set, Union
 
@@ -171,6 +172,7 @@ def recursive_model_data():
     return data
 
 
+@pytest.mark.skipif(platform.python_implementation() == 'PyPy', reason='crashes on pypy due to recursion depth')
 @skip_pydantic
 @pytest.mark.benchmark(group='recursive model')
 def test_recursive_model_pyd(recursive_model_data, benchmark):
@@ -181,6 +183,7 @@ def test_recursive_model_pyd(recursive_model_data, benchmark):
     benchmark(PydanticBranch.parse_obj, recursive_model_data)
 
 
+@pytest.mark.skipif(platform.python_implementation() == 'PyPy', reason='crashes on pypy due to recursion depth')
 @pytest.mark.benchmark(group='recursive model')
 def test_recursive_model_core(recursive_model_data, benchmark):
     class CoreBranch:

--- a/tests/test_build.py
+++ b/tests/test_build.py
@@ -63,7 +63,7 @@ def test_schema_recursive_error():
     schema = {'type': 'union', 'choices': []}
     schema['choices'].append(schema)
     with pytest.raises(
-        SchemaError, match='RecursionError: Recursive detected, depth exceeded max allowed value of 100'
+        SchemaError, match=r'RecursionError: Recursive detected, depth exceeded max allowed value of \d+'
     ):
         SchemaValidator(schema)
 

--- a/tests/validators/test_datetime.py
+++ b/tests/validators/test_datetime.py
@@ -213,7 +213,7 @@ def test_custom_invalid_tz():
     with pytest.raises(ValidationError) as excinfo:
         schema.validate_python(dt)
 
-    # exception messages are different between python and pypy
+    # exception messages differ between python and pypy
     if platform.python_implementation() == 'PyPy':
         error_message = 'NotImplementedError: tzinfo subclass must override utcoffset()'
     else:

--- a/tests/validators/test_datetime.py
+++ b/tests/validators/test_datetime.py
@@ -1,4 +1,5 @@
 import json
+import platform
 import re
 from datetime import date, datetime, time, timedelta, timezone, tzinfo
 from decimal import Decimal
@@ -211,13 +212,20 @@ def test_custom_invalid_tz():
     # perhaps this should be a ValidationError? but we don't catch other errors
     with pytest.raises(ValidationError) as excinfo:
         schema.validate_python(dt)
+
+    # exception messages are different between python and pypy
+    if platform.python_implementation() == 'PyPy':
+        error_message = 'NotImplementedError: tzinfo subclass must override utcoffset()'
+    else:
+        error_message = 'NotImplementedError: a tzinfo subclass must implement utcoffset()'
+
     assert excinfo.value.errors() == [
         {
             'kind': 'datetime_object_invalid',
             'loc': [],
-            'message': 'Invalid datetime object, got NotImplementedError: a tzinfo subclass must implement utcoffset()',
+            'message': f'Invalid datetime object, got {error_message}',
             'input_value': dt,
-            'context': {'error': 'NotImplementedError: a tzinfo subclass must implement utcoffset()'},
+            'context': {'error': error_message},
         }
     ]
 

--- a/tests/validators/test_function.py
+++ b/tests/validators/test_function.py
@@ -261,7 +261,7 @@ def test_function_wrong_sig():
 
     v = SchemaValidator({'type': 'function', 'mode': 'before', 'function': f, 'schema': {'type': 'str'}})
 
-    # exception messages are different between python and pypy
+    # exception messages differ between python and pypy
     if platform.python_implementation() == 'PyPy':
         error_message = 'f() got 2 unexpected keyword arguments'
     else:

--- a/tests/validators/test_function.py
+++ b/tests/validators/test_function.py
@@ -1,3 +1,4 @@
+import platform
 import re
 from copy import deepcopy
 
@@ -259,7 +260,14 @@ def test_function_wrong_sig():
         return input_value + ' Changed'
 
     v = SchemaValidator({'type': 'function', 'mode': 'before', 'function': f, 'schema': {'type': 'str'}})
-    with pytest.raises(TypeError, match=re.escape("f() got an unexpected keyword argument 'data'")):
+
+    # exception messages are different between python and pypy
+    if platform.python_implementation() == 'PyPy':
+        error_message = 'f() got 2 unexpected keyword arguments'
+    else:
+        error_message = "f() got an unexpected keyword argument 'data'"
+
+    with pytest.raises(TypeError, match=re.escape(error_message)):
         v.validate_python('input value')
 
 

--- a/tests/validators/test_typed_dict.py
+++ b/tests/validators/test_typed_dict.py
@@ -785,6 +785,9 @@ def test_from_attributes_error():
 
 
 def test_from_attributes_extra():
+    def another_function(x):
+        return x
+
     class Foobar:
         def __init__(self):
             self.a = 1
@@ -809,6 +812,9 @@ def test_from_attributes_extra():
         @staticmethod
         def static_method():
             return 'wrong'
+
+        # this is omitted along with the static method by the !PyFunction::is_type_of(attr) check in typed_dict.rs
+        function_attribute = another_function
 
         @classmethod
         def class_method(cls):

